kliff/trainer/kim_trainer.py

```diff
@@ -0,0 +1,181 @@
+    "Newton-CG",
+    "L-BFGS-B",
+    "TNC",
+    "COBYLA",
+    "SLSQP",
+    "trust-constr",
+    "dogleg",
+    "trust-ncg",
+    "trust-exact",
+    "trust-krylov",
+]
+
+
+class KIMTrainer(Trainer):
+    """
+    This class extends the base Trainer class for training OpenKIM physics based models.
+    It will use the scipy optimizers. It will perform a check to exclude TorchML model
+    driver based models, as they would be handled by TorchTrainer.
+    """
+
+    def __init__(self, configuration: dict, collection: str = "user"):
+        self.collection = collection
+
+        self.model_driver_name = None
+        self.parameters = None
+        self.mutable_parameters_list = []
+        self.use_energy = True
+        self.use_forces = False
+
+        super().__init__(configuration)
+
+        self.loss_function = self._get_loss_fn()
+
+    def setup_model(self):
+        """
+        Load either the KIM model, or install it from the tarball. If the model driver
+        required is TorchML* family, then it will raise an error, as it should be handled
+        by the TorchTrainer.
+        """
+        # check for unsupported model drivers
+        if (
+            self.model_manifest["type"].lower() == "kim"
+            or self.model_manifest["type"].lower() == "tar"
+        ):
+            self.model = KIMModel.get_model_from_manifest(
+                self.model_manifest, self.transform_manifest
+            )
+        else:
+            raise TrainerError(
+                f"Model type {self.model_manifest['type']} not supported."
+            )
+
+        self.parameters = self.model.get_model_params()
+
+    def setup_optimizer(self):
+        """
+        Set up the optimizer based on the provided information. If the optimizer is not
+        provided, it will be loaded from the optimizer_name. If the optimizer_name is not
+        provided, it will raise an error. If the optimizer_provider is scipy, it will be
+        loaded from the scipy.optimize. If the optimizer_provider is torch, it will be
+        loaded from the torch.optim. Left for the derived classes to implement.
+        """
+        if self.optimizer_manifest["provider"] != "scipy":
+            raise TrainerError(
+                f"Optimizer provider {self.optimizer_manifest['provider']} not supported by KIMTrainer."
+            )
+
+        if self.optimizer_manifest["name"] not in SCIPY_MINIMIZE_METHODS:
+            raise TrainerError(
+                f"Optimizer not supported: {self.optimizer_manifest['name']}."
+            )
+        optimizer_lib = importlib.import_module(f"scipy.optimize")
+        self.optimizer = getattr(optimizer_lib, "minimize")
+
+    def loss(self, x):
+        """
+        Compute the loss function for the given parameters. It will compute the loss
+        function. It seems like MPI might be only way to make it parallel as the
+        multiprocessing does not work with the KIM models. KIMPY models are not yet
```

Comment:
The current version does support parallel via multiprocess. If I remember correctly, it gets around to picking the entire KIM model object. 

But I agree -- let's focus on MPI in the future to avoid complicating it.

Reply:
That is interesting, I could not find any obvious way of pickling Pybind11 C++ bindings using pickle. Can you provide more information on it? I dont think I saw it in KIM. Just for personal interest.